src/main/java/org/broadinstitute/hellbender/tools/HaplotypeCallerSpark.java

```diff
@@ -0,0 +1,329 @@
+        final List<ReadShard.ShardBoundary> keys = intervals.stream()
+                .flatMap(interval -> ReadShard.divideIntervalIntoShards(interval, readShardSize, readShardSize, readShardPadding, header.getSequenceDictionary()).stream())
+                .collect(Collectors.toList());
+        return OverlapDetector.create(keys);
+    }
+
+    /**
+     * @return and RDD of {@link Tuple2<AssemblyRegion,SimpleInterval>} which pairs each AssemblyRegion with the
+     * interval it was generated in
+     */
+    private static FlatMapFunction<ReadShard, Tuple2<AssemblyRegion, SimpleInterval>> shardsToAssemblyRegions(
+            final AuthHolder authHolder,
+            final Broadcast<ReferenceMultiSource> reference,
+            final Broadcast<HaplotypeCallerArgumentCollection> hcArgsBroadcast,
+            final ShardingArgumentCollection assemblyArgs,
+            final SAMFileHeader header) {
+        return shard -> {
+            final ReferenceMultiSource referenceMultiSource = reference.value();
+            final ReferenceMultiSourceToReferenceDataSourceAdaptor refDataSource = new ReferenceMultiSourceToReferenceDataSourceAdaptor(referenceMultiSource, authHolder);
+            final ReferenceContext refContext = new ReferenceContext(refDataSource, shard.getPaddedInterval());
+            final ReferenceMultiSourceToReferenceSequenceFileAdaptor referenceReader = new ReferenceMultiSourceToReferenceSequenceFileAdaptor(reference.getValue(), authHolder);
+
+            //TODO load features as a side input
+            final FeatureContext features = new FeatureContext();
+
+            final Iterable<AssemblyRegion> assemblyRegions = AssemblyRegion.createFromReadShard(
+                    shard, header, refContext, features,
+                    new HaplotypeCallerEngine(hcArgsBroadcast.value(), header, referenceReader),
+                    assemblyArgs.minAssemblyRegionSize, assemblyArgs.maxAssemblyRegionSize,
+                    assemblyArgs.assemblyRegionPadding, assemblyArgs.activeProbThreshold,
+                    assemblyArgs.maxProbPropagationDistance);
+
+            return StreamSupport.stream(assemblyRegions.spliterator(), false)
+                    .map( a -> new Tuple2<>(a, shard.getInterval()))
+                    .collect(Collectors.toList());
+        };
+    }
+
+
+    @VisibleForTesting
+    static final class ReferenceMultiSourceToReferenceSequenceFileAdaptor implements ReferenceSequenceFile, Serializable{
+        private static final long serialVersionUID = 1L;
+
+        private final ReferenceMultiSource source;
+        private final AuthHolder auth;
+        private final SAMSequenceDictionary sequenceDictionary;
+
+        public ReferenceMultiSourceToReferenceSequenceFileAdaptor(final ReferenceMultiSource source, final AuthHolder auth) {
+            this.source = source;
+            this.auth = auth;
+            this.sequenceDictionary = getSequenceDictionary();
```

Comment:
calling virtual method during object initialization in constructor is bad practice